preferences for multi-root workspace

elaihau opened the pull request:
With changes in 543b119, URIs of root folders in a multi-root workspace are stored in a file with the extension of "theia-workspace", while the workspace preferences are in the ".theia/settings.json" under the first root.

In this pull request, workspace preferences are stored
- in the workspace file under the "settings" property, if a workspace file exists, or
- in the ".theia/settings.json" if the workspace data is not saved in a workspace file.

Also, this change supports
- having 4 levels of preferences (from highest priority to the lowest): FolderPreference, WorkspacePreference, UserPreference, and DefaultPreference, with
- an updated preference editor that supports viewing & editing the FolderPreferences, WorkspacePreferences, and UserPreferneces.

Signed-off-by: elaihau <liang.huang@ericsson.com>

marechal-p commented:
It seems like one of your components that is getting tested with mocha relies on some frontend package, somehow: https://travis-ci.org/theia-ide/theia/jobs/444752934#L2333

The tests run in node, so if some nested import from a chain of "require" happen to import something from phosphor indirectly, it will fail the tests, because phosphor seems to have side effects (code executed on importation) targeting a browser environment.

akosyakov commented:
> What is the status of this PR? Is it going to be merged soon?

Someone should review that there are no breaking changes in API and behavior.

lmcbout commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -236,36 +288,113 @@ export class PreferencesEditorsContainer extends DockPanel {
+            const settingsUri = await this.getFolderSettingsUri(folder);
+            const foldersPreferences = settingsUri && await this.editorManager.getOrCreateByUri(settingsUri) as PreferencesEditorWidget;
+            if (foldersPreferences) {
+                foldersPreferences.title.label = 'Folders Preferences';
```
name too long, just use the first part: i.e., User, Workspace, Folders

elaihau replied:
as discussed i will shorten texts in tab headers, and move "xxxx Preferences" into the tooltips.

elaihau commented:
@akurinnoy this pr is being reviewed by @lmcbout and others from the Ericsson team. You are also welcomed to give us any suggestions :)

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -88,10 +85,6 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
             }
         }
-        // tslint:disable-next-line:forin
-        for (const property of props) {
```
Schemas can be contributed dynamically via `setSchema`. In this case, default values should be updated.

elaihau replied:
you are right. i missed the case you described. 

as you can tell from my code, i splitted the original `PreferenceSchemaProvider` into 2 classes: the new `PreferenceSchemaProvider` (which is supposed to provide schemas) and the `DefaultPrefrenceProvider` (which is meant to provide values). I will make it work in the way that both schemas and values can be updated.

elaihau replied:
This change has been reverted.

akosyakov commented on packages/core/src/browser/preferences/preference-provider.ts:
```diff
@@ -41,20 +43,35 @@ export class PreferenceProvider implements Disposable {
+        return { priority: -1, provider: this };
+    }
+
+    getDomain(): string[] {
```
What is domain?

elaihau replied:
`domain` is where the preference changes could take effect. for example, the domain of the workspace-preferences-provider is the folders in that workspace. to be more specific, it is `["file:///A", "file:///B"]` in a workspace that has folder "A" and folder "B".

akosyakov commented on packages/core/src/browser/preferences/preference-provider.ts:
```diff
@@ -41,20 +43,35 @@ export class PreferenceProvider implements Disposable {
     }
+
+    canProvide(preferenceName: string, resourceUri?: string): { priority: number, provider: PreferenceProvider } {
+        return { priority: -1, provider: this };
```
Why does it return itself?

elaihau replied:
if you take a look at `FoldersPreferenceProvider` you would find this function does not always return the class itself. 

`UserPrefProvider.canProvide()` returns itself because there is only one `UserPrefProvider` per theia instance.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -80,12 +155,11 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
     protected init(): void {
         this.toDispose.push(Disposable.create(() => this._ready.reject()));
-        this.providers.push(this.schema);
-        this.preferences = this.parsePreferences();
```
Before it was ensured that one can query default values without awaiting on `ready`. Some clients are relying on it.

elaihau replied:
please correct me if i am wrong: before my change, clients can query default values if and only if the default schema is constructed. 

with my change, i did `defaultSchema.ready.resolve()` in the constructor (https://github.com/theia-ide/theia/pull/3247/files/7c201eacd53dfa16c259013e3ac53523f2f36af7#diff-e901dd93040742d0ffb3d98186e1043eR42). So i think that behaviour is unchanged.

akosyakov replied:
Clients can query any values immediately after the preference service is constructed. The preference service should guarantee that clients always get proper default values. It is not a case anymore because the code was moved from synchronous `postContruct` callback into async `initialize` application callback.

elaihau replied:
got it. i will move the code back to `postContruct` to make sure the prefs can be queried once the preference-service is constructed.

vince-fugnitto commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -38,18 +38,17 @@ export interface PreferenceSchema {
 
 export interface PreferenceProperty extends PreferenceItem {
     description: string;
+    scopes: PreferenceScope;
```
would `scope` be better?

lmcbout commented on packages/filesystem/src/browser/filesystem-watcher.ts:
```diff
@@ -167,16 +167,16 @@ export class FileSystemWatcher implements Disposable {
-    protected createWatchOptions(): Promise<WatchOptions> {
-        return this.getIgnored().then(ignored => ({
+    protected createWatchOptions(uri?: string): Promise<WatchOptions> {
+        return this.getIgnored(uri).then(ignored => ({
```
Why do we have the "URI" as optional when the parent calling this method is not "Optional"

elaihau commented on packages/core/src/browser/preferences/preference-provider.ts:
```diff
@@ -19,11 +19,13 @@
+import { PreferenceScope, PreferenceChange } from './preference-service';
 
 @injectable()
 export class PreferenceProvider implements Disposable {
```
Paul suggested to make this class abstract.

marechal-p replied:
Only to have `getPreferences` as abstract as well because I didn't see why we would have providers that return nothing in any situations, but I don't have a strong opinion here :)

marechal-p commented on packages/filesystem/src/browser/filesystem-preferences.ts:
```diff
@@ -35,13 +36,14 @@ export const filesystemPreferenceSchema: PreferenceSchema = {
                 '**/node_modules/**': true
-            }
+            },
+            'scopes': PreferenceScope.Default | PreferenceScope.User | PreferenceScope.Workspace | PreferenceScope.Folders
```
?

akosyakov replied:
What does VS Code do? I don't think magical numbers are not good for UI.

elaihau replied:
@akosyakov are you asking what vsCode does to define the "scopes"? 
When I looked into the vsCode source code 2 months ago, I find vsCode used hard-coded rules, such as any preference whose names start with "window" shouldn't be applied at the folder level.

elaihau replied:
vsCode has a registry that categorises all preferences into the followings:
```ts
export const allSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const applicationSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const windowSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
export const resourceSettings: { properties: {}, patternProperties: {} } = { properties: {}, patternProperties: {} };
```
From there the rest of the application knows if a preference can / cannot be applied to some certain "scope". for example, anything registered in `windowSettings` cannot be applied at the folder scope. 

in the preference schema, vsCode defined something very similar to what i have in this PR
```ts
	private registerJSONConfiguration(configuration: IConfigurationNode) {
		function register(configuration: IConfigurationNode) {
			let properties = configuration.properties;
			if (properties) {
				for (const key in properties) {
					allSettings.properties[key] = properties[key];
					switch (properties[key].scope) {
						case ConfigurationScope.APPLICATION:
							applicationSettings.properties[key] = properties[key];
							break;
						case ConfigurationScope.WINDOW:
							windowSettings.properties[key] = properties[key];
							break;
						case ConfigurationScope.RESOURCE:
							resourceSettings.properties[key] = properties[key];
							break;
					}
				}
			}
			let subNodes = configuration.allOf;
			if (subNodes) {
				subNodes.forEach(register);
			}
		}
		register(configuration);
		this._onDidSchemaChange.fire(); 
	}
```
the reason that i used `1-2-4-8` is 
- having one number to represent all 4 scopes, and the list of scopes can be extended as long as the integer does not overflow
- giving developers who add / edit  preference schema more flexibility: with my code, i can create a preference with a scope of any combination of the 4 (i.e., Default, User, Workspace, Folder), e.g., something that makes sense at the `Workspace` scope and is invalid at the `User` scope. While in vsCode, i would have choose one the three scopes: [APPLICATION, WINDOW, RESOURCE].

akosyakov replied:
I think numbers are not good for end users. It should be readable and brainer to type.

// cc @svenefftinge

akosyakov replied:
How VS Code `Application`, `Window`, `Resource` scopes are mapped to our scopes?

elaihau replied:
>I think numbers are not good for end users. It should be readable and no brainer to type.

I am using enums to avoid numbers showing up, so end users **don't** need to type any numbers. They only need to 
- not specify the `scopes` at all (in this case the `scopes` is Preference.Default + Preference.User + Preference.Workspace), or
- use any combinations of the 4 scopes that the `PreferenceScope` enum defines.

elaihau replied:
> How VS Code Application, Window, Resource scopes are mapped to our scopes?

@akosyakov 
here is the mapping between scopes defined in vsCode and this PR:
vsCode.ConfigurationScope.APPLICATION => PreferenceScope.Default + PreferenceScope.User
vsCode.ConfigurationScope.WINDOW => PreferenceScope.Default + PreferenceScope.User + PreferenceScope.Workspace
vsCode.ConfigurationScope.RESOURCE => PreferenceScope.Default + PreferenceScope.User + PreferenceScope.Workspace + PreferenceScope.Folder

With the scopes defined in vsCode, developers are unable to define a preference that is, for example, only valid at the `Workspace` and `Default` level (if i use `vsCode.ConfigurationScope.WINDOW`, users can use it at the `User` level). While with code in this PR, developers have that flexibility.

akosyakov replied:
ok

introducing similar constants would be handy, i.e. `PreferenceScope.Resource` would be handy, instead of typing `PreferenceScope.Default | PreferenceScope.User | PreferenceScope.Workspace | PreferenceScope.Folders`

elaihau replied:
sure. changed as suggested.

akosyakov replied:
We should stick to VS Code and pure JSON approach:
- otherwise, it is hard to move preferences from VS Code extensions to Theia extensions;
- If we move preferences to JSON file at some point, it won't be possible to use enums from there.

I should be called `scope` and has the same values as VS Code does.

// cc @svenefftinge

elaihau replied:
OK, I will define two enums:
```ts
export const enum PreferenceScope { 
        // this is from vsCode, supposed to be used by pref schemas
	APPLICATION = 1,
	WINDOW,
	RESOURCE,
}

export const enum PreferenceProviderScope {
   // this is the enum I already have, supposed to be used by PrefProviders internally
   Default = 1,
   User = 2,
   WS = 4,
   Folder = 8
}
```
This way we would have the compatibility with vsCode and possibility of moving to JSON files. What do you think? @akosyakov

akosyakov replied:
You should be able to type in JSON file `'scope': 'window'` and it should work, no `'scope': 1` or `'scope': 'APPLICATION'`.

Do you have a real use case for `developers are unable to define a preference that is, for example, only valid at the Workspace and Default level`? If not, please, let's think about it when it exists.

elaihau replied:
>You should be able to type in JSON file 'scope': 'window' and it should work, no 'scope': 1 or 'scope': 'APPLICATION'.

`APPLICATION = 1` is from the vsCode source code, not something I invented :) 
I personally don't think 'scope': 'window' or 'scope': 1 or 'scope': 'APPLICATION' in the json file would be a problem, we could make any (or all) of them work in future :)

>Do you have a real use case for developers are "unable to define a preference that is, for example, only valid at the Workspace and Default level"? If not, please, let's think about it when it exists.

just want to clarify: in the latest proposal described above, `PreferenceScope` and `PreferenceProviderScope` will both be defined - I wasn't promoting using one instead of the other. And only `PreferenceScope` is supposed to used in the preference schema definition. Therefore, defining `a preference that is, for example, only valid at the Workspace and Default level` is not possible any more. 

`PreferenceProviderScope` is only used internally as an enum to create PreferenceProviders. 

Speaking of `a preference that is, for example, only valid at the Workspace and Default level`, yes, I have something in mind but not sure when it will come into reality. Think about a pair programming extension that connects 2 or more developers working simultaneously on the same project (https://github.com/theia-ide/theia/issues/1405). Different developers, in many cases, prefer to have different settings, e.g., a pref that toggles the highlight of the line that the other developer is working on (google docs has the highlight = on by default). With that "hightlight_the_other_guy = on", developers have better feel of collaboration and would less likely to step on each other's toes when editing the same package or file. imo, to have "hightlight_the_other_guy = on", we would need to add a new scope, which changes the PreferenceProviderScope into

```ts
export const enum PreferenceProviderScope {
   // this is the enum I already have, supposed to be used by PrefProviders internally
   Default = 1,
   User = 2,
   WS = 4,
   Folder = 8,
   PerInstance = 16 // or called InMemory
}
```
and it does not make sense to have that "hightlight_the_other_guy = on" defined in a scope other than `PerInstance`.

elaihau replied:
@akosyakov hi Anton, are you ok with my proposal described above (i.e., having PreferenceScope defined in vsCode to achieve the compatibility, and PreferenceProviderScope for provider creation) ? I prefer to get your opinion before making this change, as it would change a lot of lines. Thanks !

akosyakov replied:
@elaihau my concern is to be compatible with VS code because of reasons I've put [above](https://github.com/theia-ide/theia/pull/3247#discussion_r237105884). If it would be possible then go ahead and do what you think is necessary.

elaihau replied:
@akosyakov i updated the values used in `PreferenceScope` to keep them aligned with those used in vsCode.

vince-fugnitto commented:
@elaihau as we mentioned, I think having the `scope` optional in the schemas would be better due to extenders and not break the current api. If a preference scope is not provided by extenders then we can treat their scope as either `default` or `user` level.

lmcbout commented on packages/navigator/src/browser/navigator-filter.ts:
```diff
@@ -14,7 +14,7 @@
 
-import { inject, injectable } from 'inversify';
+import { inject, injectable, postConstruct } from 'inversify';
 import { Minimatch } from 'minimatch';
```
Module 'minimatch' is not listed as dependency in package.json

elaihau replied:
i will address this comment in a separate PR :)

elaihau commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -60,22 +59,20 @@ export function bindPreferenceSchemaProvider(bind: interfaces.Bind): void {
 
 @injectable()
-export class PreferenceSchemaProvider extends PreferenceProvider {
+export class PreferenceSchemaProvider {
```
i reverted this change - `PreferenceSchemaProvider` now extends the `PreferenceProvider` that implements `Disposable`.

lmcbout commented on packages/preferences/src/browser/abstract-resource-preference-provider.ts:
```diff
@@ -54,14 +55,15 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
         this.toDispose.push(resource);
         if (resource.onDidChangeContents) {
-            this.toDispose.push(resource.onDidChangeContents(content => this.readPreferences()));
+            this.onDidResourceChanged = resource.onDidChangeContents(content => this.readPreferences());
```
variable "content" is never used, do we need it ?

lmcbout commented on packages/preferences/src/browser/folder-preference-provider.ts:
```diff
@@ -0,0 +1,73 @@
+            return super.canProvide(preferenceName, resourceUri);
+        }
+        const uri = new URI(resourceUri);
+        return { priority: 3 + this.folderUri.path.relativity(uri.path), provider: this };
```
priority value should be defined in a central location, apply to user and workspace as well

elaihau replied:
yes, i will add a "PreferenceProviderPriority" namespace to preference-provider.ts.

lmcbout commented on packages/preferences/src/browser/folders-preferences-provider.ts:
```diff
@@ -0,0 +1,140 @@
+        if (this.workspaceService.saved) {
+            for (const root of this.workspaceService.tryGetRoots()) {
+                if (await this.fileSystem.exists(root.uri)) {
+                    const provider = await this.createFolderPreferenceProvider(root);
```
Invalid 'await' of a non-Promise value

lmcbout commented on packages/preferences/src/browser/folders-preferences-provider.ts:
```diff
@@ -0,0 +1,140 @@
+        // becomes ready even if we fail reading the preferences, so we don't
+        // hang the preference service.
+        Promise.all(this.providers.map(p => p.ready))
+            .then(() => this._ready.resolve())
```
As per coding guidelines, https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names
we should not use the "_" for private variables

elaihau commented on packages/preferences/src/browser/folders-preferences-provider.ts:
```diff
@@ -0,0 +1,140 @@
+        Promise.all(this.providers.map(p => p.ready))
+            .then(() => this._ready.resolve())
+            .catch(() => this._ready.resolve());
+
```
it would be the best if we could do 
```ts
Promise.all(...blah...)
  .finally(() => this._ready.resolve());
```
but `Promise.finally` is unavailable in this project for now.

The reason that we want to do `ready.resolve()` in both cases is that, we want to unblock the clients from consume the data as long as all providers finish loading. The loading being successful or not successful is not a concern here, as we don't want any features in Theia to hang up if a preference value is (temporarily) inaccessible.

lmcbout commented on packages/preferences/src/browser/preference-service.spec.ts:
```diff
@@ -54,11 +55,10 @@ disableJSDOM();
-const mockUserPreferenceEmitter = new Emitter<void>();
-const mockWorkspacePreferenceEmitter = new Emitter<void>();
+const mockUserPreferenceEmitter = new Emitter<PreferenceChange>();
+const mockWorkspacePreferenceEmitter = new Emitter<PreferenceChange>();
```
Could we mock the folderPreference and a a test on it as well ?

lmcbout commented on packages/preferences/src/browser/preference-service.spec.ts:
```diff
@@ -68,16 +68,22 @@ before(async () => {
+                    mockUserPreferenceEmitter.event
+                );
+                return userProvider;
+            default:
```
If we add workspaceScope and FolderScope, we could add a test for anything else and see if ever we try to modify the scope functionality, what do you think ?

lmcbout commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -212,19 +253,31 @@ export class PreferencesEditorsContainer extends DockPanel {
+    private _userPreferenceEditorWidget: PreferencesEditorWidget;
+    private _workspacePreferenceEditorWidget: PreferencesEditorWidget | undefined;
+    private _foldersPreferenceEditorWidget: PreferencesEditorWidget | undefined;
 
```
For the tree private variables (_userPreferenceEditorWidget, _workspacePreferenceEditorWidget and _foldersPreferenceEditorWidget),
If we follow the Theia coding guide-lines, we should not use the "_" in front of the variables 
( https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names )

lmcbout commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -236,36 +289,116 @@ export class PreferencesEditorsContainer extends DockPanel {
-        userPreferences.title.label = 'User Preferences';
-        userPreferences.title.caption = await this.getPreferenceEditorCaption(userPreferenceUri);
+        userPreferences.title.label = 'User';
+        userPreferences.title.caption = `User Preferneces: ${await this.getPreferenceEditorCaption(userPreferenceUri)}`;
```
typo: Preferneces !! -> Preferences

lmcbout commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -236,36 +289,116 @@ export class PreferencesEditorsContainer extends DockPanel {
+            workspacePreferences.title.label = 'Workspace';
+            workspacePreferences.title.caption = `Workspace Preferences: ${await this.getPreferenceEditorCaption(workspacePreferenceUri!)}`;
+            workspacePreferences.title.icon = 'database-icon medium-yellow file-icon';
+            workspacePreferences.editor.setLanguage(JSONC_LANGUAGE_ID);
```
We add the "language" setting  here for the workspace editor, but we didn't add it for the user and the folder workspaces

elaihau replied:
it is because the user and folder settings files use `.json` file extension - it can be recognised by theia, while workspace settings file (for now) could possibly have any extension. We don't force users to use `theia-workspace` or `code-workspace` - maybe we should ?

lmcbout commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -236,36 +289,116 @@ export class PreferencesEditorsContainer extends DockPanel {
+                await this.fileSystem.createFile(settingsUri);
+            }
+            return new URI(settingsUri);
+        }
```
Creating a second time the URI  "settingsUri",  why not using the one created above and use "toString()" when needed a few lines above

elaihau replied:
if i did not misunderstand this comment, you suggested to use the "settingsUri" in line 371.

i believe the current code is already doing what you suggested. line 371 calls `getFolderSettingsUri()`, and uses the returns value after.  the "settingsUri" is created once only.

elaihau replied:
we talked in person. you are right. i will make the requested change.

lmcbout commented on packages/workspace/src/browser/workspace-frontend-contribution.spec.ts:
```diff
@@ -37,31 +37,24 @@ describe('workspace-frontend-contribution', () => {
-            [OS.Type.Linux, 'electron', true, { title, canSelectFiles: true, canSelectFolders: false, filters }],
-            [OS.Type.Linux, 'electron', false, { title, canSelectFiles: false, canSelectFolders: true }],
+            [OS.Type.Linux, 'browser', { title, canSelectFiles: true, canSelectFolders: true, filters }],
+            [OS.Type.Linux, 'electron', { title, canSelectFiles: true, canSelectFolders: false, filters }],
```
Question: any reason why some tests were removed ?

elaihau replied:
i will revert this change - as per the short discussion we had in today's dev meeting, the workspace.supportMultiRoot would stay for now. So there is no need to remove tests.

lmcbout commented on packages/workspace/src/browser/workspace-frontend-contribution.ts:
```diff
@@ -359,21 +333,23 @@ export namespace WorkspaceFrontendContribution {
         }
 
-        // Otherwise, it is always a folder.
+        // In electron, only workspace files can be selected when the multi-root workspace feature is enabled.
```
is mentioning "multi-root" here is necessary ? since we remove the flag

elaihau replied:
the flag will come back. so i will keep it as is.

lmcbout commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -115,16 +123,33 @@ export class WorkspaceService implements FrontendApplicationContribution {
     }
 
     protected async updateWorkspace(): Promise<void> {
+        if (this._workspace) {
```
As per Theia coding guide lines, we should not use "_" in front of the private variable
 i.e.  "_workspace"
ref: https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names

marechal-p replied:
Collision with https://github.com/theia-ide/theia/blob/ce1e855e8702dc009383c8630bafc45354e68df2/packages/workspace/src/browser/workspace-service.ts#L97

marechal-p replied:
It is common to have `_something` when you have a `something` getter in the same class.

elaihau commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -236,36 +289,116 @@ export class PreferencesEditorsContainer extends DockPanel {
         const userPreferences = await this.editorManager.getOrCreateByUri(userPreferenceUri) as PreferencesEditorWidget;
-        userPreferences.title.label = 'User Preferences';
-        userPreferences.title.caption = await this.getPreferenceEditorCaption(userPreferenceUri);
+        userPreferences.title.label = 'User';
```
we can but we don't need to, as the `.json` extension can be recognised by theia, and the same icon would be added automatically. 

this comment helps me find out 
```ts
foldersPreferences.title.icon = 'database-icon medium-yellow file-icon';
```
is redundant. I will remove it. thank you!

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -59,36 +59,45 @@ export function bindPreferenceSchemaProvider(bind: interfaces.Bind): void {
+        for (const property of Object.keys(schema.properties)) {
+            const schemaProps = schema.properties[property];
             if (this.combinedSchema.properties[property]) {
-                console.error('Preference name collision detected in the schema for property: ' + property);
```
`console.error` always calls `Logger` under the hood, there is no need to use it explicitly

akosyakov commented on packages/cpp/src/browser/cpp-preferences.ts:
```diff
@@ -50,11 +50,13 @@ export const cppPreferencesSchema: PreferenceSchema = {
             default: false,
-            type: 'boolean'
+            type: 'boolean',
+            scopes: PreferenceScope.Default & PreferenceScope.User & PreferenceScope.Workspace
```
Could you make `scopes` optional, use `PreferenceScope.Default & PreferenceScope.User & PreferenceScope.Workspace` as default, revert changes to all preferences, only apply custom `scopes` where it is necessary? Otherwise it is a breaking change.

elaihau replied:
you are right ! in the latest version of code i made it optional.

akosyakov replied:
What is the default?

elaihau commented:
i made the preference scope optional in order to achieve the backward compatibility.
Also, as per the suggestion from @vince-fugnitto , i updated the readme.md in the preferences package.

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -59,13 +59,18 @@ export function bindPreferenceSchemaProvider(bind: interfaces.Bind): void {
+    protected readonly combinedSchema: PreferenceSchema = { properties: {} };
     protected validateFunction: Ajv.ValidateFunction;
 
+    @inject(ILogger) protected readonly logger: ILogger;
```
console always using `ILogger`, using `ILogger` is just more code

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -74,21 +79,25 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
     protected doSetSchema(schema: PreferenceSchema): void {
         const props: string[] = [];
-        for (const property in schema.properties) {
+        for (const property of Object.keys(schema.properties)) {
```
`in` operator is an equivalent of `of Object.keys`, just shorter

elaihau replied:
i got an eslint error `for (... in ...) statements must be filtered with an if statement (forin)`  when `for...in` was used.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -43,29 +69,78 @@ export interface PreferenceChanges {
+    protected readonly onPreferenceChangedEmitter = new Emitter<PreferenceDataChange>();
     readonly onPreferenceChanged = this.onPreferenceChangedEmitter.event;
 
     protected readonly onPreferencesChangedEmitter = new Emitter<PreferenceChanges>();
```
it does not seem to be used, but there are client using it, please fire events properly

elaihau replied:
you are right. i removed this event and updated the client.

akosyakov commented on packages/core/src/browser/preferences/preference-provider.ts:
```diff
@@ -14,16 +14,24 @@
+export abstract class PreferenceProvider implements Disposable {
+
+    protected readonly onDidPreferencesChangedEmitter = new Emitter<PreferenceChange>();
+    readonly onDidPreferencesChanged: Event<PreferenceChange> = this.onDidPreferencesChangedEmitter.event;
```
it is breaking change, it used to be composite event, to indicate that there are some preferences changed, not one like it is now

elaihau replied:
yes i changed the data that is emitted with the Event, but i am not sure why it is a breaking change:

- PreferenceProvider is used internally. Clients only use PreferenceService.
- Even if there are clients listening to `PreferenceProvider.onDidPreferencesChanged`, the clients' code is not broken by this change, because the clients' code does not expect getting anything from the event.

akosyakov replied:
`MonacoConfiguration` should completely reconcile all preferences on a change:  https://github.com/theia-ide/theia/blob/6a3444d341b4e16048a18f3588d0249f33d8f57c/packages/monaco/src/browser/monaco-configurations.ts#L43

Before it was not for each preference name, but for composite. You changed make it very expensive now, instead of reconciling once, it reconciles for each preference name.

Please put back `onPreferencesChangedEmitter` or rewrite `MonacoConfiguration` more efficient without breaking it.

akosyakov commented on packages/core/src/browser/preferences/preference-proxy.ts:
```diff
@@ -17,29 +17,37 @@
+    readonly preferenceName: keyof T;
+    readonly newValue?: T[keyof T];
+    readonly oldValue?: T[keyof T];
+    canAffect(resourceUri?: string): boolean;
```
why `can`? Should not it tell whether it is affected or not? just `isAffected` or `affects`?

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -222,4 +280,26 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
 
+    inspect<T>(preferenceName: string, resourceUri?: string): {
+        preferenceName: string,
+        values: Map<PreferenceScope, T>
```
each object is JS is already a map, following API like would be nicer and aligned with VS Code:
```ts
{
     preferenceName: string,
     defaultValue: T,
     globalValue: T, // USER
     workspaceValue: T,
     workspaceFolderValue: T,
}
```

elaihau replied:
i updated this function to match the vscode API

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -16,25 +16,51 @@
 export interface PreferenceChange {
     readonly preferenceName: string;
+    readonly scope: PreferenceScope
+    readonly domain: string[];
```
it is an implementation detail, end users should not care, they already have `scope` and a function `affects` to check whether their resource is affected by a change

We should hide `domain` as implementation details.

elaihau replied:
i reverted my changes to `PreferenceChange`.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,77 +175,48 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
+        return prefs;
     }
 
     has(preferenceName: string): boolean {
```
it should also accept `resourceUri?: string` as the second arg.

elaihau commented:
@akosyakov 

Hi Anton, I noticed that you added java-preferences where "scope" is used.

i guess your "scope" does not mean the same thing with the "PreferenceScope" in my PR, does it ?

In this case, Could you please suggest a different property name that I could use in my PR? I am asking because your comment suggested me to use "scope" as it is used in the vsCode the other day. 

Thank you !

elaihau commented:
I did more investigation on how preference scopes are defined and schemas contribution works in vsCode.

Internally, vsCode uses the following enum to define the preference scope:
```typescript
export const enum ConfigurationScope {
	APPLICATION = 1,
	WINDOW,
	RESOURCE,
}
```
For example, the schemas defined in [search.contribution.ts](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/search/electron-browser/search.contribution.ts#L605):
```typescript
'search.useIgnoreFiles': {
   type: 'boolean',
   markdownDescription: nls.localize('useIgnoreFiles', "Controls whether to use `.gitignore` and `.ignore` files when searching for files."),
   default: true,
   scope: ConfigurationScope.RESOURCE
},
```

External extensions, however, make contributions to preferences by having schemas defined in the package.json, e.g., [vscode-java's package.json](https://github.com/redhat-developer/vscode-java/blob/master/package.json):
```json
    "configuration": {
      "type": "object",
      "title": "Java configuration",
      "properties": {
        "java.home": {
          "type": [
            "string",
            "null"
          ],
          "default": null,
          "description": "...",
          "scope": "window"
        }
      }
   }
```
And vsCode converts the preference scope from strings to enums with the [following code](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/configuration/common/configurationExtensionPoint.ts#L176):
```typescript
if (propertyConfiguration.scope) {
	if (propertyConfiguration.scope.toString() === 'application') {
		propertyConfiguration.scope = ConfigurationScope.APPLICATION;
    	} else if (propertyConfiguration.scope.toString() === 'resource') {
		propertyConfiguration.scope = ConfigurationScope.RESOURCE;
	} else {
		propertyConfiguration.scope = ConfigurationScope.WINDOW;
	}
} else {
	propertyConfiguration.scope = ConfigurationScope.WINDOW;
}
```

From my point of view, we could leave schemas defined in [java-preferences.ts (theia)](https://github.com/theia-ide/theia/blob/master/packages/java/src/browser/java-preferences.ts) as is, and perform the string=>enum conversion in [PreferenceSchemaProvider](https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/preferences/preference-contribution.ts#L63)

@akosyakov

akosyakov commented:
@elaihau yes, we need to support both formats, since enum representation is public API as well: https://code.visualstudio.com/api/references/vscode-api#ConfigurationTarget

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -102,22 +135,36 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
+        for (const property of Object.keys(schema.properties)) {
+            const schemaProps = schema.properties[property];
+            this.emitPreferenceChangedEvent(property, schemaProps.default, undefined);
+        }
```
@elaihau here new changes to `PreferencesFrontendApplicationContribution`: https://github.com/theia-ide/theia/pull/4003/files#diff-9edbd6a2baf11d0ba53136cf40cb9ab7R40

elaihau replied:
`this.emitPreferenceChangedEvent()` in this change calls `this.onDidPreferencesChangedEmitter.fire()`, so it wouldn't break PR 4003.

akosyakov replied:
It breaks because before reconciliation was happening only once for all properties not for each. It is expensive now. You set a schema with 3 new preferences and instead of serializing the combined scheme once, it would be serialized 3 times.

elaihau replied:
Got it. I reverted my changes to `PreferenceServiceImpl.reconcilePreferences()`, and bring back `onPreferencesChanged`.

In the current version, I have both `onPreferenceChanged` for every single settings change, and `onPreferencesChanged` for clients that need bulk updates.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
-            for (const provider of providers) {
-                provider.onDidPreferencesChanged(_ => this.reconcilePreferences());
+            for (const [scope, provider] of this.providersMap.entries()) {
+                if (scope !== PreferenceScope.DEFAULT) {
```
Why default is filtered out? they can change as well then someone contributes schema dynamically

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
             }
-            this.reconcilePreferences();
-            this._ready.resolve();
+            Promise.all(this.providers.map(p => p.ready)).then(() => this._ready.resolve());
```
what about `this._ready` when one of `p.ready` throws?

elaihau replied:
in the current code (both on master and this PR), p.ready from the preference-providers [never rejects](https://github.com/theia-ide/theia/blob/master/packages/preferences/src/browser/abstract-resource-preference-provider.ts#L50), and therefore we don't take care of the scenario you are referring to.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
-                    changes[preferenceName] = { preferenceName, newValue, oldValue };
-                    this.preferences[preferenceName] = deepFreeze(newValue);
+
+    protected createProviders(): void {
```
fyi: Are you sure that there are not subclasses overriding these method outside of Theia repo?

We should be careful doing changes to APIs to avoid causing issues like: https://github.com/theia-ide/theia/issues/4034#issuecomment-454099723

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
+        );
+    }
+
+    protected reconcilePreferences(changes: PreferenceChanges): void {
```
it looks suspicious now.

Before all scopes were taken into the account, for example, if preferences changed on user level, but it is already overriden on workspace then there should not be any events.
Now it seems to care about only one scope. It is bogus, should be like before.

elaihau replied:
it cares about all scopes, the logic is between lines 162-174

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -100,119 +124,132 @@ export class PreferenceServiceImpl implements PreferenceService, FrontendApplica
-        for (const preferenceName in changes) {
-            this.onPreferenceChangedEmitter.fire(changes[preferenceName]);
+        if (Object.keys(changesToEmit).length > 0) {
+            this.onPreferencesChangedEmitter.fire(changesToEmit);
```
before it would compute proper composite change event, filter out irrelevant changes, like not it just dump raw changes

elaihau replied:
it does compute proper composite change event, the logic is between lines 155-174

akosyakov commented on packages/editor/src/browser/editor-preferences.ts:
```diff
@@ -32,12 +32,14 @@ export const editorPreferenceSchema: PreferenceSchema = {
             'default': 4,
-            'description': 'Configure the tab size in the editor.'
+            'description': 'Configure the tab size in the editor.',
+            'scope': 'resource'
```
VS Code allows to define default scope for schema: https://github.com/Microsoft/vscode/blob/6b8a54df78848c1fa582557fc4ca583d67177a56/src/vs/editor/common/config/commonEditorConfig.ts#L227

Could we do the same instead of copy & paste?

elaihau replied:
Yes. Done in the latest verison.

akosyakov commented on packages/core/src/browser/frontend-application-module.ts:
```diff
@@ -188,8 +188,10 @@ export const frontendApplicationModule = new ContainerModule((bind, unbind, isBo
-    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.User);
-    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.Workspace);
+    bindDefaultPreferenceProvider(bind);
+    bind(PreferenceProvider).toSelf().inSingletonScope().whenTargetNamed(PreferenceScope.USER);
```
renaming enum literals is breaking change, are you sure there are not extension relying on it. We agreed to not break if we are not sure or go through deprecation: https://github.com/theia-ide/theia/wiki/Dev-Meetings#api-deprecation

akosyakov commented:
The content assist does not work in workspace settings file:
<img width="2032" alt="screen shot 2019-01-17 at 08 31 30" src="https://user-images.githubusercontent.com/3082655/51302364-682d5880-1a32-11e9-9e99-91dcc3fb716f.png">

It would be enough to file a follow-up issue.

akosyakov commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -16,27 +16,67 @@
     readonly preferenceName: string;
     readonly newValue?: any;
     readonly oldValue?: any;
+    readonly scope: PreferenceScope;
```
How can i check whether a change affected me in some context?

I remember you had something like `affects(resourceUri?: string): boolean`.

I don't understand how i can use  `scope`  to check affection. Maybe  `scope` also should not be on an event?

elaihau replied:
removed `scope` and added `affects()` back. Thank you for the review !

akosyakov commented:
As a follow-up `MonacoConfiguration` should be rewritten to support multiple roots. @elaihau please open an issue

akosyakov commented:
I've tested without multi-root workspaces and it seems to work like before. With multi-root workspaces there are weird behaviour in `PreferenceService` clients, for example in `MonacoConfiguration`. They receive changes about all folders, but does not care about them.

@svenefftinge could you test it as well? 

There are a lot of incompatible changes, so I guess we need the major release for sure if we merge it. Unfortunately without deprecation. @marcdumais-work

elaihau commented:
> The content assist does not work in workspace settings file:
> It would be enough to file a follow-up issue.

Thank you for the test ! could you please kindly advise how `new URI('vscode://schemas/settings/user')` is translated into the preference schema? To fix the bug you found, i will need to registered a new schema with 
```ts
jsonSchemaStore.registerSchema({
   fileMatch: [ ... , ...],
   url
});
``` 
where the `url` points to the new schema.

elaihau commented:
> There are a lot of incompatible changes

@akosyakov @marcdumais-work 

i don't have much experiences in making backward-compatible API changes, so please correct me if i am wrong :)

i made the following interface / api changes:
1) changed what PreferenceProvider emits on a change being made to `settings.json`. The PreferenceProvider used to emit an event with no data. In this PR, the providers are expected to emit data that tells the service what was changed, what are the old & new values, etc.

This could introduce compatibility issues, if extension developers replace theia preferences extension with some preferences-customised.

2) added the `affects()` function to `PreferenceChange` interface.
3) added one optional argument to `PreferenceService.get()` and `PreferenceService.set()`

I don't think 2) or 3) are breaking changes. Since having folder-level preferences isn't supported in master branch, no extensions / plugins should contribute uri-sensitive preferences to theia. Therefore, the clients could continue using the old way to access data from the event and service.

4) added `PreferenceService.inspect()`

An added function shouldn't be a breaking change.

maybe i have missed something ?

akosyakov commented:
Changing methods which are not private, i.e. renaming, changing return types, removing arguments, adding new not optional arguments are breaking changes, e.g. simple like `fireOnDidPreferencesChanged` -> `emitPreferencesChangedEvent`. There could be implementations of `PreferenceProvider` and they would be broken.

I don't know though that anyone provides custom `PreferenceProvider` or `PreferenceService`, so it could be fine that we ignore it this time or just release the major version without deprecation. 

// cc @svenefftinge @marcdumais-work 

> i don't have much experiences in making backward-compatible API changes, so please correct me if i am wrong :)

Usually, one implements new APIs, deprecates old without changing signatures and after several releases remove old. Sometimes one reimplements old APIs with new to remove unnecessary code. 

We are building the framework that allows fine-grained customizations, but with that maintainers have to care about backward-compatible changes, i.e. do minimal change, don't change not private APIs unnecessary (because you are annoyed by typo or `emit` seems to be better naming than `fire` for you), if you do, do it controllable via deprecation.

elaihau commented:
> Changing methods which are not private, i.e. renaming, changing return types, removing arguments, adding new not optional arguments are breaking changes, e.g. simple like `fireOnDidPreferencesChanged` -> `emitPreferencesChangedEvent`. There could be implementations of `PreferenceProvider` and they would be broken.
> 
> I don't know though that anyone provides custom `PreferenceProvider` or `PreferenceService`, so it could be fine that we ignore it this time or just release the major version without deprecation.

Thank you for the explanation. Really appreciated !

I talked with @marcdumais-work and will try make a version that supports both old & new ways of handling pref changes, with deprecation of course.

elaihau commented:
@akosyakov I made more changes made to the API in PreferenceProvider & PreferenceService to achieve the backward compatibility. Could you please take a look when you get the chance ? Thank you !

akosyakov commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -59,11 +92,15 @@ export function bindPreferenceSchemaProvider(bind: interfaces.Bind): void {
 }
 
+export function bindDefaultPreferenceProvider(bind: interfaces.Bind): void {
+    bind(PreferenceProvider).to(PreferenceSchemaProvider).inSingletonScope().whenTargetNamed(PreferenceScope.Default);
```
It creates a completely new instance of `PreferenceSchemaProvider`, you should use `toService` or `toDynamicValue` to get an instance of `PreferenceSchemaProvider` bound by `bindPreferenceSchemaProvider`.

elaihau replied:
Thank you for the review, i went through my changes and realized that `bindDefaultPreferenceProvider` is no longer needed. `PreferenceSchemaProvider` is now only injected, and never created by `PreferenceProviderProvider`. So this function is removed.

akosyakov commented on packages/preferences/src/browser/abstract-resource-preference-provider.ts:
```diff
@@ -15,22 +15,23 @@
     // tslint:disable-next-line:no-any
     protected preferences: { [key: string]: any } = {};
+    protected resource: Promise<Resource>;
+    protected onDidResourceChanged: Disposable | undefined;
```
why do you need this property?

elaihau replied:
workspace preferences are stored
1) in the ".theia/settings.json" if the workspace data is not saved in a workspace file, or
2) in the workspace file under the "settings" property, if a workspace file exists.

when the state transitions from #1 to the #2, WorkspacePreferenceProvider is not interested in changes made to ".theia/settings.json" any more (an instance of FolderPreferenceProvider will take care of it instead), and therefore, i want to dispose `onDidResourceChanged` when the state transition happens.

akosyakov commented on packages/preferences/src/browser/abstract-resource-preference-provider.ts:
```diff
@@ -54,14 +55,15 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
         this.toDispose.push(resource);
         if (resource.onDidChangeContents) {
-            this.toDispose.push(resource.onDidChangeContents(content => this.readPreferences()));
+            this.onDidResourceChanged = resource.onDidChangeContents(() => this.readPreferences());
```
I would create a new `toDisposeOnInit` disposable collection and dispose it on `init` call, in subclass i would only call ìnit`.

elaihau replied:
sure I can add `this.resource` and `this.onDidResourceChanged` to a collection named `toDisposeOnWorkspaceLocationChanged`.

akosyakov commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -117,6 +125,11 @@ export class WorkspaceService implements FrontendApplicationContribution {
     }
 
+    protected readonly onSavedLocationChangeEmitter = new Emitter<FileStat | undefined>();
+    get onSavedLocationChanged(): Event<FileStat | undefined> {
```
`onWorkspaceLocationChaned`

akosyakov commented:
browser console full or errors if try to edit setting files:
<img width="1922" alt="screen shot 2019-01-28 at 09 53 34" src="https://user-images.githubusercontent.com/3082655/51824485-b1a75e80-22e2-11e9-82ba-28b6f0c4b3f3.png">

Reproducible by adding `files.watcherExclude` into workspace preferences.

elaihau commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -59,11 +92,15 @@ export function bindPreferenceSchemaProvider(bind: interfaces.Bind): void {
     bindContributionProvider(bind, PreferenceContribution);
 }
 
+export function bindDefaultPreferenceProvider(bind: interfaces.Bind): void {
```
like i mentioned above this function has been removed.

elaihau commented:
resolved conflicts between this PR and https://github.com/theia-ide/theia/commit/6b1a35cf249dc8c6c73cdf058cba4c5673b7e31c

elaihau commented:
@akosyakov I noticed that you bumped the version to 0.4.x. Can I also take advantage of the version change and merge this PR in near future ?

akosyakov commented:
@elaihau i've looked at the code, thank you for the effort, not it has much less incompatible changes than an original version. It is easier to follow without being afraid that something broken. 

Unfortunately, they are still new mandatory arguments, so yes the major version is required. Could you add a record in `Change Log` under `Breaking Changes`? I would test tomorrow again and hopefully we can merge afterwards.

akosyakov commented on CHANGELOG.md:
```diff
@@ -15,6 +15,10 @@ Breaking changes:
     - `3_move` and `7_actions` replaced by `navigation` group
   - editor context menu group changes:
     - `2_cut_copy_paste` renamed to `9_cutcopypaste` group
+- [core] `PreferenceProvider` [#3247](https://github.com/theia-ide/theia/pull/3247)
```
i would group them under `multi-root workspace support for preferences`. There are more incompatible changes to what is listed here.

elaihau replied:
i updated the list. could you check if there are items missing?

akosyakov commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -117,6 +125,11 @@ export class WorkspaceService implements FrontendApplicationContribution {
         return this.onWorkspaceChangeEmitter.event;
     }
 
+    protected readonly onSavedLocationChangeEmitter = new Emitter<FileStat | undefined>();
```
`onWorkspaceLocationChangedEmitter`

akosyakov commented:
It seems to behave properly. @svenefftinge Can you test it as well?

elaihau commented:
@benoitf @evidolob could you please check if this PR is ok for plugin developers ? Thank you !

elaihau commented:
ping @evidolob and @benoitf  one more time for approval :)
